### DISCUSSION TITLE
Clarify semantics of app context

### DISCUSSION THREAD

**Discussion by wwwillchen** [pallets/flask#5507] [Q&A] [answered]
Thanks for building Flask! I'm one of the maintainers of https://github.com/google/mesop which is a Python UI framework built on Flask. Mesop uses `g` to maintain request-specific context (essentially the UI data for the current request), see [call site example](https://github.com/google/mesop/blob/ba2e889aaa694f2223244c51734151c9544e9b6c/mesop/runtime/runtime.py#L63)

I read through the Flask docs and am not 100% sure whether this is an appropriate use of `g`.

From https://flask.palletsprojects.com/en/3.0.x/api/#flask.g:

> This is a good place to store resources during a request. For example, a before_request function could load a user object from a session id, then set g.user to be used in the view function.

This makes it sound like `g` is suitable as I'm trying to store information about the current request.

---

From https://flask.palletsprojects.com/en/3.0.x/appcontext/:
> Flask automatically pushes an application context when handling a request

This suggests that an application context is _always_ created inside the scope of a request.

However, https://flask.palletsprojects.com/en/3.0.x/api/#flask.ctx.AppContext states:
> An app context is created and pushed at the beginning of each request if one is not already active

This seems to suggest an application context is _not_ always created inside the scope of a request. These two docs seem to contradict one another (or perhaps I'm mis-reading it). Any clarification would be helpful.

---

One of the complexities with our use case is that other developers are building on top of Mesop and could create app contexts, and I want to make sure that when I'm binding a value to `g`, it's to a local-specific app context and not some more global app context that persists across requests.

**Comment by davidism** [comment, marked as answer]
> This makes it sound like `g` is suitable as I'm trying to store information about the current request.

`g` is the appropriate place to store per-request data. For extensions, we recommend using a prefix like `_extension_name_` as a namespace to avoid any potential name conflicts. See https://flask.palletsprojects.com/en/3.0.x/extensiondev/ and specifically https://flask.palletsprojects.com/en/3.0.x/extensiondev/#data-during-a-request for extension dev docs about this.

**Reply by davidism** [reply]
The app context and request context work side by side. They are separate because there are times when you'll be using the app outside of requests, such as during setup or in CLI commands. During requests, there should not already be an application context, so it's created along with the request context. If there is already and application context, that suggests that someone incorrectly pushed a context and didn't pop it, so you shouldn't worry about that case.

**Reply by wwwillchen** [reply]
got it, thanks for the answer! I'll follow your suggestion with the naming.
